test/TestTemplate.hs

```diff
@@ -0,0 +1,92 @@
+module TestTemplate (tests) where
+
+import Data.Either (isLeft, isRight)
+import Data.Either.Combinators (leftToMaybe, rightToMaybe)
+import Data.Maybe (fromMaybe)
+import Data.Text (intercalate)
+import Data.Text qualified as T
+import Data.Text.Arbitrary ()
+import Lhx (Separator (Separator))
+import Lhx qualified as L
+import Lhx.Parser qualified as LP
+import Test.Tasty (TestTree, testGroup)
+import Test.Tasty.HUnit (assertEqual, testCase, (@?), (@?=))
+import Test.Tasty.QuickCheck qualified as QS
+
+makeTemplateTests :: TestTree
+makeTemplateTests =
+  testGroup
+    "make template"
+    [ testGroup
+        "fail"
+        [ testCase "parse errors remain" $
+            isLeft (L.makeTemplate "$foo") @? "parsing error should remain"
+        , testCase "unknown function" $
+            isLeft (L.makeTemplate "$foo;") @? "function isn't in the list functions"
+        , testCase "all unknown functions must be caught" $
+            leftToMaybe (L.makeTemplate "$foo:bar:rev:baz:rev:bazzz:rev;")
+              @?= Just
+                [ L.Error "Unknown function: foo"
+                , L.Error "Unknown function: bar"
+                , L.Error "Unknown function: baz"
+                , L.Error "Unknown function: bazzz"
+                ]
+        ]
+    , testGroup
+        "right"
+        [ testCase "all good functions must be parsed" $
+            isRight
+              ( L.makeTemplate $
+                  "$" <> intercalate ":" (map (LP.unFName . fst) L.functions) <> ";"
+              )
+              @? "don't apply function in the list functions"
+        ]
+    ]
+
+makeInputTest :: TestTree
+makeInputTest =
+  testGroup
+    "make input"
+    [ testCase "empty" $ L.makeInput (L.Separator " ") "" @?= L.Input "" [""]
+    , testCase "separator ','" $
+        L.makeInput (L.Separator ",") "a, b, c" @?= L.Input "a, b, c" ["a", " b", " c"]
+    ]
+
+applyFunctionTest :: TestTree
+applyFunctionTest =
+  let reverseTemplate = L.makeTemplate "$rev;"
+   in testGroup
+        "Apply tests"
+        [ QS.testProperty "reverse function" $
+            \s -> fromMaybe False $
+              rightToMaybe $ do
+                template <- reverseTemplate
+                text <- L.apply template (L.makeInput (Separator ",") s)
+                pure $ text == T.reverse s
```

Comment:
It is a good taste to have all the properties declared at the top level. Or in `where` at least.